Feature/bind as user

olikasg opened the pull request:
###  Expose LDAP ports in testing compose
commit 86ab66f6e2351294e1c263c99d3761ba0942c1e3   

    Expose the LDAP ports in the docker-compose used for CI. This way, the
    LDAP container is usable as a standalone testing server and useful for
    manual testing.

### Introduce new LDAP configuration options
commit 41cd1547c95f6a0332fc18c31e6a9b77b7fcde8f
    
    `auth.ldap.custom_base_dn` can be used as a substitute for the
    hard-coded base DN used for both authentication and authorization. The
    hardcoded version mandated to have the user from the connection to exist
    in LDAP under the `uid=` attribute. The new base DN allows for fully
    custom string. Three variables can be used and they will be substituted
    before sending the DN to LDAP.
      - `${username_attr}` will be substitured to the value defined in
      `auth.ldap.username.attributetype`.
      - `${user}` is substituted to the username provided by the client.
      - `${device_dn}` is substituted to the string defined in
      `auth.ldap.device_dn`
    
    Additionaly a typo was fixed in the `auth.ldap.device_dn` default
    string.
    
    `auth.ldap.filters.$num.{key|value|op}` can be used to specify
    additional filter values compared to the predefined
    `objectClass=Class`. At least the `key` and `value` must be present and
    they represent an equality check. At each level (`$num`) an optional
    operation can be defined (currently only "and" or "or"). This applies to
    all the previously defined filters in a left-associative manner. There
    is no way to express hierarchy. An example can be found in a comment in
    the schema file. The default option is the old behaviour if no filter is
    present in the config.
    
    `auth.ldap.bind_as_user` must be set to `true` if a second bind
    operation is required to authenticate the user provided in the
    client. Otherwise the user will only used for authorization. If this
    option is omitted or set to `false`, then the old behaviour is
    used.

### Add support for "Bind as User" feature
commit 0339b835823db463b359dcedaa82ab1450f0ad4a
    
    In some cases, the Base DN to authenticate the user provided by the
    connection needs to be obtained by an LDAP search. This change makes it
    possible to do so in EMQx.
    
    The substitution of variables in the configuration paramters are
    performed unconditionally for both the user and acl processing. This was
    necessary to remove the hardcoded Base DN generation and still preserve
    backward compatibility. All users benefit from this small improvement,
    regardless whether they need the "Bind As User" feature.
    
    The substitution processing is not optimised, but the lists are expected
    to be very small or empty, thus no major performance degradation is
    expected.
    
    In the LDAP cli implementation, an extra bind is introduced, if
    `bind_as_user` is set to true. This was necessary to reset the last bind
    to the root user before executing the query.
    
    A major consideration was taken to preserve backward compatibility for
    users not opting for the feature. No configuration changes are
    necessary, but the `priv/eqmx_auth_ldap.schema` needs to be updated to
    pick up the proper default values.

### Add additional LDAP schema attribute
commit 120162eab02845db9e95075661875f581574896c
    
    Add an extra schema attribute to the LDAP schema of EMQx. This was used
    for testing and could be removed, if it interferes with policies.

olikasg committed:
Expose LDAP ports in testing compose

Expose the LDAP ports in the docker-compose used for CI. This way, the
LDAP container is usable as a standalone testing server and useful for
manual testing.

olikasg committed:
Introduce new LDAP configuration options

`auth.ldap.custom_base_dn` can be used as a substitute for the
hard-coded base DN used for both authentication and authorization. The
hardcoded version mandated to have the user from the connection to exist
in LDAP under the `uid=` attribute. The new base DN allows for fully
custom string. Three variables can be used and they will be substituted
before sending the DN to LDAP.
  - `${username_attr}` will be substitured to the value defined in
  `auth.ldap.username.attributetype`.
  - `${user}` is substituted to the username provided by the client.
  - `${device_dn}` is substituted to the string defined in
  `auth.ldap.device_dn`

Additionaly a typo was fixed in the `auth.ldap.device_dn` default
string.

`auth.ldap.filters.$num.{key|value|op}` can be used to specify
additional filter values compared to the predefined
`objectClass=Class`. At least the `key` and `value` must be present and
they represent an equality check. At each level (`$num`) an optional
operation can be defined (currently only "and" or "or"). This applies to
all the previously defined filters in a left-associative manner. There
is no way to express hierarchy. An example can be found in a comment in
the schema file. The default option is the old behaviour if no filter is
present in the config.

`auth.ldap.bind_as_user` must be set to `true` if a second bind
operation is required to authenticate the user provided in the
client. Otherwise the user will only used for authorization. If this
option is omitted or set to `false`, then the old behaviour is
used.

olikasg committed:
Add support for "Bind as User" feature

In some cases, the Base DN to authenticate the user provided by the
connection needs to be obtained by an LDAP search. This change makes it
possible to do so in EMQx.

The substitution of variables in the configuration paramters are
performed unconditionally for both the user and acl processing. This was
necessary to remove the hardcoded Base DN generation and still preserve
backward compatibility. All users benefit from this small improvement,
regardless whether they need the "Bind As User" feature.

The substitution processing is not optimised, but the lists are expected
to be very small or empty, thus no major performance degradation is
expected.

In the LDAP cli implementation, an extra bind is introduced, if
`bind_as_user` is set to true. This was necessary to reset the last bind
to the root user before executing the query.

A major consideration was taken to preserve backward compatibility for
users not opting for the feature. No configuration changes are
necessary, but the `priv/eqmx_auth_ldap.schema` needs to be updated to
pick up the proper default values.

olikasg committed:
Add additional LDAP schema attribute

Add an extra schema attribute to the LDAP schema of EMQx. This was used
for testing and could be removed, if it interferes with policies.

luos commented on src/emqx_auth_ldap.erl:
```diff
@@ -62,14 +64,67 @@ check(ClientInfo = #{username := Username, password := Password}, AuthResult,
+        end,
+
+    %% auth.ldap.custom_base_dn = "${username_attr}=${user},${device_dn}"
+    BaseDN = replace_vars(CustomBaseDN, ReplaceRules),
```
Would it worth moving these parts into their own function as it looks pretty similar to the `acl` part. (Line 73-98)

luos commented on src/emqx_auth_ldap.erl:
```diff
@@ -62,14 +64,67 @@ check(ClientInfo = #{username := Username, password := Password}, AuthResult,
+                    {"${device_dn}", DeviceDn}],
+    PasswordString = binary_to_list(Password),
+
+    %% Here the original filter could be appended to the list of filters, like
```
I am not quite sure what this comment refers to. 
Is it required to append? Or just could be? Seemingly below it's never appended (Line 93), only used if there were no other rules.

olikasg replied:
The comment was removed. The old behaviour used this filter by default. With the new mechanism, if no filter is defined, then the old behaviour is applied. If any filter is defined, not it is the operators responsibility to do the same.

luos commented on src/emqx_auth_ldap_cli.erl:
```diff
@@ -63,24 +65,74 @@ connect(Opts) ->
-                             end).
+    ecpool:with_client(?APP,
+        fun(C) ->
+                case application:get_env(?APP, bind_as_user) of
```
I think here these two `search/2,3` could be merged into one, if there is no significant difference?  I am not sure if there is an acceptable default value for `attributes`.

olikasg replied:
They are used in different circumstances. `search/3` is used for ACL's, the other is for users. Keeping the original design for now, as it is not covered by tests at all.

luos commented on priv/emqx_auth_ldap.schema:
```diff
@@ -89,6 +89,70 @@ end}.
+%%   [{"objectClass","mqttUser"},"and",{"uiAttr","someAttr"},"or",{"someKey","someAttr"}]
+%% The resulting LDAP filter would look like this:
+%% ==> "|(&(objectClass=Class)(uiAttr=someAttr)(someKey=someValue))"
+{translation, "emqx_auth_ldap.filters",
```
Do think it would be possible to add some simple validation to `or` and `and` `ops` so it is a good error message if something else is provided? 

Is there some format requirement for the attr `key` or `values`?

olikasg replied:
There is a validation is the source code. Any other value and there will be a function clause exception. I found that cuttlefish reporting on bad parameters is a bit unreliable. Sometimes I didn't get any indication why the node didn't start. So I'd rather leave it as is. Furthermore, this kind of configuration is cumbersome (to say the least). I would rather push for adopting the RMQ's LDAP plugin althogether.

luos commented on priv/emqx_auth_ldap.schema:
```diff
@@ -89,6 +89,70 @@ end}.
+    {datatype, string}
+]}.
+
+{mapping, "auth.ldap.filters.filters.$num.op", "emqx_auth_ldap.filters", [
```
Seems like `{datatype, {enum, [ "or", "and" ] } }` would fit here instead of `string`.

olikasg replied:
Thanks. Will change.

velimir commented on src/emqx_acl_ldap.erl:
```diff
@@ -32,6 +32,9 @@
 
 -import(emqx_auth_ldap_cli, [search/3]).
 
+-import(emqx_auth_ldap, [replace_vars/2,
```
I would rather reduce usage of import statements as they make code harder to read.

olikasg replied:
Will change after the current implementation is verified.

velimir commented on src/emqx_auth_ldap.erl:
```diff
@@ -62,14 +64,67 @@ check(ClientInfo = #{username := Username, password := Password}, AuthResult,
+                            {error, Reason}
+                    end;
+                [Val] ->
+                    case list_to_atom(string:to_lower(Val)) of
```
would be nice to replace it to `list_to_existing_atom/1` at some point

velimir replied:
I know that it was inspired by the previous implementation. Just a note, I think it would be great to catch an error  if a value can't be casted to a boolean.

olikasg replied:
Agreed. The value that can be passed is bound by the LDAP schema:
```
attributetype ( 1.3.6.1.4.1.11.2.53.2.2.3.1.2.3.1.3 NAME 'isEnabled'
	EQUALITY booleanMatch
```
I will leave it as is for now, as it is the old behaviour.

velimir commented on src/emqx_auth_ldap_cli.erl:
```diff
@@ -63,24 +65,74 @@ connect(Opts) ->
                 error:Reason -> {error, Reason}
             end;
-        {error, Reason} ->
+        {_, {error, Reason}} ->
```
Why was this changed?

olikasg replied:
Ooops. Remnant of a reverted change.

olikasg replied:
Removed the change.

velimir commented on src/emqx_auth_ldap_cli.erl:
```diff
@@ -63,24 +65,74 @@ connect(Opts) ->
+                end
+        end).
+
+post_bind(BindDn, BindPassword) ->
```
function name is confusing. This is not a `post_bind`, it's just a bind. The context in which the function is called is a "post" operation.

olikasg replied:
Hm... I will change the name. If it is only called a bind, then it might be confusing why it is not used in the previous functions. This module would be a good candidate for a bigger refactoring after this change. The question is when.

olikasg replied:
I'll leave it for now. When this whole module is refactored to cater for a more generic configuration then it can be changed.

velimir commented on src/emqx_auth_ldap.erl:
```diff
@@ -133,3 +188,35 @@ do_resolve(HashType, Passhash, Password) ->
 
 description() -> "LDAP Authentication Plugin".
 
+compile_filters([{Key, Value}], []) ->
```
I assume it just works, but would be great to see unit/functional tests for this later on.

olikasg replied:
Agreed. It will come later, if the current implementation works.

olikasg replied:
I added a config that exercises this logic.

olikasg replied:
Since I don't think this filter declaration is a long-lasting solution, I'd rather not spend too much time writing tests.

velimir reviewed (commented):
some comments, plus agree with @luos comments.

olikasg committed:
Add tests to cover the new functionality

olikasg committed:
Refactoring to abstract repeated code

Based on review comments I moved the common filter processing logic to
its own function.

Also eliminating newly introduced imports.

olikasg committed:
Change the operator config type definition

olikasg committed:
Rework the post-bind call

If bind-as-user is active, then there is no need to check the password
from the attributes.

olikasg committed:
Add missing attribute to the LDAP test data

olikasg committed:
Fix config schema for the op parameter
